0015-estimator-interface.md

```diff
@@ -0,0 +1,457 @@
+
+```python
+class Estimator(BasePrimitive):
+
+    # current signature
+    def run(
+            self,
+            circuits: Sequence[QuantumCircuit] | QuantumCircuit,
+            observables: Sequence[BaseOperator | PauliSumOp | str] | BaseOperator | PauliSumOp | str,
+            parameter_values: Sequence[Sequence[float]] | Sequence[float] | float | None = None,
+            **run_options,
+        ):
+        ...
+```
+
+We propose the migration to using tasks in two phases.
+
+### Deprecation Phase 1
+
+In the first phase:
+* Introduce `tasks: Sequence[EstimatorTask] | EstimatorTask` as the only positional argument, and make `circuits` a keyword argument.
+* Coerce the arguments to account for the fact that the only existing positional argument is `circuit: Sequence[QuantumCircuit] | QuantumCircuit`.
+* Check that the user does not attempt to mix the old/new APIs.
+* If necessary, coerce the old-API arguments (now all keyword arguments) into `EstimatorTask`s, **raise deprecation warning**.
+* Always eventually run with `Estimator._run(tasks: Sequence[EstimatorTask], **run_options)`.
+
+Here is a mock implementation, with missing type annotations as above:
+```python
+class Estimator(BasePrimitive):
+
+    def run(
+        self, 
+        tasks: Sequence[EstimatorTask] | EstimatorTask, 
+        circuits = None,
+        observables = None, 
+        parameter_values = None, 
+        **run_options
+    ):
+        # Coerce into standard form to later allow for `tasks` to be the only non-kwarg.
+        if isinstance(tasks, QuantumCircuit) or all(isinstance(task, QuantumCircuit) for task in tasks):
+            if circuits is not None:
+                raise ValueError("Cannot mix old and new APIs")
+            circuits = tasks
+            tasks = None
+
+        # Do not mix new/old API
+        if tasks and (circuits or observables or parameter_values):
+            raise ValueError("Cannot mix old and new APIs")
+
+        # Deprecated path
+        if tasks is None:
+            # Verify values in old API
+            if not (circuits and observables):
+                raise ValueError(f"Need both circuits and observables for old API, got circuits={circuits}, observables={observables}")
+            
+            # Coerce into old form
+            if parameter_values is None:
+                parameter_values = itertools.repeat(None)
+            if isinstance(circuit, QuantumCircuit):
+                circuit = [circuit]
+            if isinstance(observables, (BaseOperator, PauliSumOp, str)):
+                observables = [observables]
+
+            # Coerce old form into `EstimatorTask`s
+            tasks = [EstimatorTask.coerce(task) for task in zip(circuits, observables, parameter_values)]
```

Comment:
It would be good to add a validate step of numbers of circuits, observables, and parameter_values before generating tasks. Otherwise, the number of tasks could be smaller than expected e.g. if the number of observables is smaller than that of circuits.

Reply:
Would you be able to write a patch or code suggestion that I can commit?

Reply:
(bearing in mind that the main purpose of this block is to convince everyone that there is a feasible migration path, so I'd avoid adding it in a way that introduces clutter)

Reply:
I see. Validation of inputs is not the focus of this RFC. So, I'm fine to skip it.